akka-http-core/src/main/scala/akka/http/engine/client/HostConnectionPoolSlot.scala

```diff
@@ -0,0 +1,187 @@
+  sealed trait SlotEvent extends ProcessorOut
+  sealed trait SimpleSlotEvent extends SlotEvent
+  object SlotEvent {
+    final case class RequestCompleted(slotIx: Int) extends SimpleSlotEvent
+    final case class Disconnected(slotIx: Int, failedRequests: Int) extends SimpleSlotEvent
+    final case class RetryRequest(rc: RequestContext) extends SlotEvent
+  }
+
+  case class Ports(
+    requestContextIn: Inlet[RequestContext],
+    responseOut: Outlet[ResponseContext],
+    slotEventOut: Outlet[SlotEvent]) extends Shape {
+
+    override val inlets = requestContextIn :: Nil
+    override def outlets = responseOut :: slotEventOut :: Nil
+
+    override def deepCopy(): Shape = Ports(
+      new Inlet(requestContextIn.toString),
+      new Outlet(responseOut.toString),
+      new Outlet(slotEventOut.toString))
+
+    override def copyFromPorts(inlets: immutable.Seq[Inlet[_]], outlets: immutable.Seq[Outlet[_]]): Shape =
+      Ports(
+        inlets.head.asInstanceOf[Inlet[RequestContext]],
+        outlets.head.asInstanceOf[Outlet[ResponseContext]],
+        outlets.last.asInstanceOf[Outlet[SlotEvent]])
+  }
+
+  private val slotProcessorActorName = new SeqActorName("SlotProcessor")
+
+  /*
+    Stream Setup
+    ============
+
+    Request-   +-----------+                       +-------------+                +------------+ 
+    Context    | Slot-     |  List[ProcessorOut]   |   flatten   |  ProcessorOut  | SlotEvent- |    ReponseContext
+    +--------->| Processor +---------------------->| (MapConcat) +--------------->| Split      +------------------>
+               |           |                       |             |                |            |                                    
+               +-----------+                       +-------------+                +-----+------+                                    
+                                                                                        | SlotEvent                                                    
+                                                                                        | (to Conductor
+                                                                                        |  via slotEventMerge)
+                                                                                        v 
+   */
+  def apply(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any],
+            remoteAddress: InetSocketAddress, // TODO: remove after #16168 is cleared
+            settings: HostConnectionPoolSettings)(implicit system: ActorSystem, fm: FlowMaterializer): Graph[Ports, Unit] =
+    FlowGraph.partial() { implicit b ⇒
+      import FlowGraph.Implicits._
+
+      val slotProcessor = b.add {
+        Flow[RequestContext] andThenMat { () ⇒
+          val actor = system.actorOf(Props(new SlotProcessor(slotIx, connectionFlow, settings)), slotProcessorActorName.next())
+          ActorProcessor[RequestContext, List[ProcessorOut]](actor) -> ()
```

Comment:
Okay, this is the one that needs a bit more thought, I feel that I’m not far from coming up with a good idea, hold on to your hats!